## Statistics Demonstrations


- Distributions (Binomial and Normal)
- Random Numbers
- Combinations

In [1]:
#Importing the libraries
import numpy as np
#Introducing the scipy.stats library
import scipy.stats as ss

### Distributions

##### 1. Binomial Distribution

A `binomial distribution` is utilised when the following conditions are met

- Total number of trials is fixed at **n**

- Each trial is binary, i.e., has only two possible outcomes - success or failure

- Probability of success is same in all trials, denoted by **p**

Now if you want to find the probability of **r** successes or `P(X = r)`, then you can use the following formula

![Binomial Probability Formula](images\binom.jpg)

In [2]:
#Calculating probabilites
#P (X = r)
n = 10
p = 0.4
r = 4

In [3]:
#To calculate the probability we'll use the binom.pmf
?ss.binom.pmf

In [4]:
ss.binom.pmf(r,n,p)

0.2508226560000002

In [5]:
#round the values
np.round(ss.binom.pmf(r,n,p),3)

0.251

In [6]:
#Calculating Probabilites
# P(X < = r)
#Use the binom.cdf function Cumulative Distribution function - cdf
?ss.binom.cdf

In [7]:
#Calculating the probabilities
np.round(ss.binom.cdf(r,n,p),3)

0.633

###### 2. Normal Distribution

![Normal Distribution](images\normd.jpg)

Let's say you're given a normal distribution with

- mean = M
- standard deviation = sd

Now you are given a value **x** and you want to compute P(X<=**x**)

In [8]:
M = 60
sd = 10
x = 40

In [9]:
#To compute P(X<=x) we use the ss.norm.cdf function
?ss.norm.cdf

#### 1st method

In [10]:
ss.norm.cdf(x,M,sd)

0.022750131948179195

##### 2nd Method

In [11]:
ss.norm(60,10).cdf(40)

0.022750131948179195

Calculating the probability of value lies between 40 to 80 with mean value = 60 and standard deviation = 10

In [12]:
ss.norm.cdf(80,60,10) - ss.norm.cdf(40,60,10)

0.9544997361036416

As per 1-2-3 rule of normal distribution, we can say that the value lies between Mean+2(sigma) Mean-2(Sigma). (60+2*10,60-2*10) = (80,40). From the rule we can say the value is 95.4%. Our answer is also the same value.

### Random Numbers

In [13]:
#Generate random numbers belonging to a particular distribution

In [14]:
?np.random

##### 1. Binomial Distribution

For binomial distribution, given the value of **n** (the number of trials) and **p** ( the probability of success for each trial) we can generate random sequence of  possible number of successes **r** when we run this experiment multiple times.

In [15]:
?np.random.binomial

In [16]:
#Generate 10 random numbers from the binomial distribution with n = 10 and p = 0.4 
n = 10
p = 0.4
size = 10

Generating the possible number of success values r

In [17]:
#We'll use np.random.binomial
np.random.binomial(n,p,size)

array([3, 3, 4, 4, 5, 3, 4, 4, 4, 4])

###### 2. Normal Distribution

For normal distribution, given the value of mean (**M**) and standard deviation (**sd**) we can generate random sequence of values belonging to this distribution

In [18]:
?np.random.normal

In [19]:
#Generate 10 random numbers from the normal distribution with M = 60 and sd = 10
M = 60
sd  = 10

Generating the random variable sequence(X) using random function

In [20]:
np.random.normal(M,sd,10)

array([51.14845797, 63.10332985, 69.6323002 , 65.38804621, 68.90114331,
       49.4934592 , 69.25110181, 50.07063521, 61.44461039, 58.78011978])

The values generated by random will change everytime we run. So get the fixed values we use seed.


###### Seed

In order to fix the output we can set the `seed` value

In [21]:
?np.random.seed

In [22]:
seed = 100
np.random.seed(seed)

In [23]:
a = np.random.normal(M,sd,10)
print(a)

[42.50234527 63.42680403 71.53035803 57.47563963 69.81320787 65.14218841
 62.21179669 49.29956669 58.10504169 62.55001444]


In [24]:
np.random.normal(M,sd,10)

array([55.41973014, 64.35163488, 54.1640495 , 68.16847072, 66.72720806,
       58.95588857, 54.68719623, 70.29732685, 55.61864377, 48.81681754])

The seed has to be initialised each time you run the code to get the same output

### Combinations

Given a list containing **n** number of objects, we can find all the combinations of size **r**

In [25]:
from itertools import combinations

In [26]:
?combinations

In [41]:
alpha = 'ABCD'

In [42]:
p = combinations(alpha,2)

In [43]:
p

In [44]:
#To print the combinations
list(p)

[('A', 'B'), ('A', 'C'), ('A', 'D'), ('B', 'C'), ('B', 'D'), ('C', 'D')]

In [45]:
alpha = ['A','B','C','D']

In [46]:
p = combinations(alpha,2)

In [47]:
a = list(p)
print(a)

[('A', 'B'), ('A', 'C'), ('A', 'D'), ('B', 'C'), ('B', 'D'), ('C', 'D')]


In [50]:
len(a)

6

In [35]:
num = [1,2,3]
num

[1, 2, 3]

In [36]:
list(combinations(num,2))

[(1, 2), (1, 3), (2, 3)]

In [37]:
for c in combinations(num,2):
    if 2 in c:
        print(c)

(1, 2)
(2, 3)


In [38]:
for c in combinations(alpha,2):
    if 'A' in c:
        print(c)

('A', 'B')
('A', 'C')
('A', 'D')
